judge/judgeapi.py

```diff
@@ -100,6 +104,65 @@ def judge_submission(submission, rejudge=False, batch_rejudge=False, judge_id=No
     return success
 
 
+def batch_judge_submission(submissions, rejudge=False, judge_id=None):
+    from .models import Submission, SubmissionTestCase
+    updates = {
+        'time': None, 'memory': None, 'points': None, 'result': None, 'case_points': 0, 'case_total': 0,
+        'error': None, 'rejudged_date': timezone.now() if rejudge else None, 'status': 'QU',
+        'is_pretested': Coalesce(
+            Subquery(
+                Submission.objects.filter(pk=OuterRef('pk'))
+                .annotate(pretested=F('contest_object__run_pretests_only').bitand(F('contest__problem__is_pretested')))
+                .values_list('pretested')[:1],
+                output_field=BooleanField(),
+            ),
+            False,
+        ),
+    }
+
+    with transaction.atomic():
+        submission_queryset = Submission.objects.filter(id__in=map(attrgetter('id'), submissions)) \
+                                                .exclude(status__in=('P', 'G'))
+        submission_queryset.update(**updates)
+
+        ids = set(submission_queryset.values_list('id', flat=True))
+        # Do the filtering using the new set of IDs rather than submission_queryset
+        # because submission_queryset itself takes a list of IDs anyways.
+        SubmissionTestCase.objects.filter(submission_id__in=ids).delete()
+
+    submissions = [submission for submission in submissions if submission.id in ids]
+
+    try:
+        response = judge_request({
+            'name': 'batch-submission-request',
+            'submissions': [{
+                'submission-id': submission.id,
+                'problem-id': submission.problem.code,
+                'language': submission.language.key,
+                'source': submission.source.source,
+            } for submission in submissions],
+            'judge-id': judge_id,
+            # This is technically incorrect when rejudge is False.
+            # Some submissions might be contest submissions and should have priority "CONTEST_SUBMISSION_PRIORITY".
+            # It's not worth further complicating this code by separating contest submissions and
+            # normal submissions as that case is not even being used.
+            # For the moment, let's just set all submissions to have "DEFAULT_PRIORITY" if we're not rejudging.
+            'priority': BATCH_REJUDGE_PRIORITY if rejudge else DEFAULT_PRIORITY,
```

Comment:
Why not just move the `priority` into the list? Then you can set per-submission priorities. In fact priority is a per-submission priority and should be in there anyway...